aten/src/ATen/native/cuda/MixedDtypesLinear.cu

```diff
@@ -257,40 +260,82 @@ _mixed_dtypes_linear(const Tensor& input, const Tensor& weight,
   TORCH_CHECK(input_2d.size(1) == weight.size(0),
```

Comment:
btw this error check needs to be removed for 4-bit.

Reply:
This check there is OK, the trick is in the interface to this functionality.  Namely, for performance reasons, underlying CUTLASS-based implementation (reused from FasterTransformer project) of mixed dtypes GEMM requires that elements of quantized matrix be reordered up-front.  So, as shown by `test_mixed_dtypes_linear` test (from `test/test_matmul_cuda.py`), at the moment usage of this functionality of PyTorch would go as follows:
```
from torch.quantization._quantized_conversions import (
    quantized_weight_reorder_for_mixed_dtypes_linear_cutlass,
)

dtypeq = torch.int8 # or torch.quint4x2
weightq = ... # of type dtypeq
scale = ...

weigthq_reordered = quantized_weight_reorder_for_mixed_dtypes_linear_cutlass(
    weightq, dtypeq, transpose=True
)

# ...

input1 = ...
output1 = torch.ops.aten._mixed_dtypes_linear(
   input1, weightq_reordered, scale, bias
)

# ...

input2 = ...
output2 = torch.ops.aten._mixed_dtypes_linear(
   input2, weightq_reordered, scale, bias
)

# and so on
```

So the `weight` tensor that is an argument in the C++ code is not as usual for linear operator (on the other side, having `scale` argument there is also unusual), and the checks are written according to what the reordering method mentioned above returns.  Maybe it would be better if the `weight` argument is named differently, or maybe if all of this wrapped through sub-classing of tensor class.  But there is [ongoing work](https://github.com/NVIDIA/cutlass/pull/1084) to have mixed dtypes GEMM implemented directly in CUTLASS upstream, that seems to avoid reordering the quantized matrix, so hopefully this complication will be removed from this PyTorch code soon.

Reply:
I guess I'm not sure how that resolves the issue. At the end of the day, `weight`'s size is now mismatched with the `input` size, no?

And this check would error.

Concretely, I was unable to run your benchmarking script without making this change (and also some changes to transposes).

Reply:
Ah, sorry - the comment with the benchmarking script was not updated for the latest changes, now it is.

Let's say the input matrix is `n` by `k`, the weight matrix for ordinary linear operator would be `m` by `k`, but here it would be passed to the reordering routine with the `transpose=True` argument, and it would return `k` by `m` reordered matrix.  Thus, the check in C++ code is fine.